kubetest/kubeadmdind/kubeadm_dind.go

```diff
@@ -0,0 +1,567 @@
+		if err := d.execCmdSaveLog(cmder, cmd, logDir, logFile); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// clusterSize determines the number of nodes in a cluster.
+func (d *Deployer) clusterSize() (int, error) {
+	o, err := d.getOutput("kubectl get nodes --no-headers")
+	if err != nil {
+		return -1, fmt.Errorf("kubectl get nodes failed: %s\n%s", err, string(o))
+	}
+	trimmed := strings.TrimSpace(string(o))
+	if trimmed != "" {
+		return len(strings.Split(trimmed, "\n")), nil
+	}
+	return 0, nil
+}
+
+// Create a local log artifacts directory
+func (d *Deployer) makeLogDir(logDir string) error {
+	cmd := fmt.Sprintf("mkdir -p %s", logDir)
+	execCmd := d.execCmd(cmd)
+	return d.control.FinishRunning(execCmd)
+}
+
+// saveHostLogs collects service logs and docker state from the host
+// container, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveHostLogs(artifactsDir string) error {
+	log.Printf("Saving logs from host container")
+
+	// Create directory for the host container artifacts
+	logDir := filepath.Join(artifactsDir, "host-container")
+	if err := d.run("mkdir -p " + logDir); err != nil {
+		return err
+	}
+
+	// Save docker state for the host container
+	if err := d.saveDockerState(d.hostCmder, logDir); err != nil {
+		return err
+	}
+
+	// Copy service logs from the node container
+	return d.saveServiceLogs(d.hostCmder, hostServices, logDir)
+}
+
+// saveMasterNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs from all nested master node containers that are running
+// on the host container, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveMasterNodeLogs(artifactsDir string) error {
+	masters, err := d.detectNodeContainers(kubeMasterPrefix)
+	if err != nil {
+		return err
+	}
+	for _, master := range masters {
+		if err := d.saveNodeLogs(master, artifactsDir, systemdServices, masterKubePods); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// saveWorkerNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs from all nested worker node containers that are running
+// on the host container, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveWorkerNodeLogs(artifactsDir string) error {
+	nodes, err := d.detectNodeContainers(kubeNodePrefix)
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if err := d.saveNodeLogs(node, artifactsDir, systemdServices, nodeKubePods); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// detectNodeContainers creates a list of names for either all master or all
+// worker node containers. It does this by running 'kubectl get nodes ... '
+// and searching for container names that begin with a specified name prefix.
+func (d *Deployer) detectNodeContainers(namePrefix string) ([]string, error) {
+	log.Printf("Looking for container names beginning with '%s'", namePrefix)
+	o, err := d.getOutput("kubectl get nodes --no-headers")
```

Comment:
I'm curious to know where does it the the kubeconfig from at this point?

Reply:
The $HOME/.kube/config file for kubectl gets created in the host DinD container via a 'kubectl config set-cluster ...' command. The K-D-C scripts do that e.g. here: https://github.com/kubernetes-sigs/kubeadm-dind-cluster/blob/master/dind-cluster.sh#L1091-L1093. (I think that's what you were asking).